app/src/main/java/com/duckduckgo/app/sitepermissions/SitePermissionsViewModel.kt

```diff
@@ -0,0 +1,170 @@
+                    _viewState.emit(
+                        _viewState.value.copy(
+                            sitesPermissionsAllowed = sitePermissionsList,
+                            locationPermissionsAllowed = locationPermissionsList
+                        )
+                    )
+                }
+            }
+        }
+    }
+
+    fun permissionToggleSelected(
+        isChecked: Boolean,
+        textRes: Int
+    ) {
+        when (textRes) {
+            R.string.sitePermissionsSettingsLocation -> {
+                settingsDataStore.appLocationPermission = isChecked
+                _viewState.value = _viewState.value.copy(askLocationEnabled = isChecked)
+                removeLocationSites()
+                fireTogglePixel(isChecked, PixelValue.LOCATION)
+            }
+            R.string.sitePermissionsSettingsCamera -> {
+                sitePermissionsRepository.askCameraEnabled = isChecked
+                _viewState.value = _viewState.value.copy(askCameraEnabled = isChecked)
+                fireTogglePixel(isChecked, PixelValue.CAMERA)
+            }
+            R.string.sitePermissionsSettingsMicrophone -> {
+                sitePermissionsRepository.askMicEnabled = isChecked
+                _viewState.value = _viewState.value.copy(askMicEnabled = isChecked)
+                fireTogglePixel(isChecked, PixelValue.MIC)
+            }
+        }
+    }
+
+    private fun removeLocationSites() {
+        viewModelScope.launch {
+            geolocationPermissions.clearAll()
+            _viewState.emit(_viewState.value.copy(locationPermissionsAllowed = listOf()))
+        }
+    }
+
+    private fun fireTogglePixel(enabled: Boolean, paramValue: String) {
+        val pixelName = when (enabled) {
+            true -> SitePermissionsPixelName.SITE_PERMISSIONS_ASK_ENABLED
+            false -> SitePermissionsPixelName.SITE_PERMISSIONS_ASK_DISABLED
+        }
+        pixel.fire(pixelName, mapOf(PixelParameter.SITE_PERMISSION to paramValue))
+    }
+
+    fun allowedSiteSelected(domain: String) {
+        viewModelScope.launch {
+            _commands.send(LaunchWebsiteAllowed(domain))
+        }
+    }
+
+    fun removeAllSitesSelected() {
+        val sitePermissions = _viewState.value.sitesPermissionsAllowed.toMutableList()
+        val locationPermissions = _viewState.value.locationPermissionsAllowed.toMutableList()
+        viewModelScope.launch(dispatcherProvider.io()) {
+            geolocationPermissions.clearAll()
+            sitePermissionsRepository.deleteAll()
+            _commands.send(ShowRemovedAllConfirmationSnackbar(sitePermissions, locationPermissions))
+        }
+    }
+
+    fun onSnackBarUndoRemoveAllWebsites(
+        removedSitePermissions: List<SitePermissionsEntity>,
+        removedLocationPermissions: List<LocationPermissionEntity>
+    ) {
+        viewModelScope.launch(dispatcherProvider.io()) {
+            sitePermissionsRepository.undoDeleteAll(removedSitePermissions)
```

Comment:
Same for this one. have you considered putting this inside a permissions repository instead of having this in the viewmodel?

Reply:
Yes, it will be moved to SitePermissionsRepository when Location is migrated and updated with new UI